command/plan.go

```diff
@@ -18,112 +16,173 @@ type PlanCommand struct {
-		return 1
-	}
+	// Parse and validate flags
+	args, diags := arguments.ParsePlan(rawArgs)
 
-	configPath, err := ModulePath(cmdFlags.Args())
-	if err != nil {
-		c.Ui.Error(err.Error())
+	// Instantiate the view, even if there are flag errors, so that we render
+	// diagnostics according to the desired view
+	view := views.NewPlan(args.ViewType, c.RunningInAutomation, c.View)
+
+	if diags.HasErrors() {
+		view.Diagnostics(diags)
+		view.HelpPrompt("plan")
 		return 1
 	}
 
 	// Check for user-supplied plugin path
+	var err error
 	if c.pluginPath, err = c.loadPluginPath(); err != nil {
-		c.Ui.Error(fmt.Sprintf("Error loading plugin path: %s", err))
+		diags = diags.Append(err)
+		view.Diagnostics(diags)
 		return 1
 	}
 
-	var backendConfig *configs.Backend
-	var configDiags tfdiags.Diagnostics
-	backendConfig, configDiags = c.loadBackendConfig(configPath)
-	diags = diags.Append(configDiags)
-	if configDiags.HasErrors() {
-		c.showDiagnostics(diags)
+	// FIXME: the -input flag value is needed to initialize the backend and the
+	// operation, but there is no clear path to pass this value down, so we
+	// continue to mutate the Meta object state for now.
+	c.Meta.input = args.InputEnabled
+
+	// FIXME: the -parallelism flag is used to control the concurrency of
+	// Terraform operations. At the moment, this value is used both to
+	// initialize the backend via the ContextOpts field inside CLIOpts, and to
+	// set a largely unused field on the Operation request. Again, there is no
+	// clear path to pass this value down, so we continue to mutate the Meta
+	// object state for now.
+	c.Meta.parallelism = args.Operation.Parallelism
+
+	diags = diags.Append(c.providerDevOverrideRuntimeWarnings())
+
+	// Prepare the backend with the backend-specific arguments
+	be, beDiags := c.PrepareBackend(args.State)
+	diags = diags.Append(beDiags)
+	if diags.HasErrors() {
+		view.Diagnostics(diags)
 		return 1
 	}
 
-	// Load the backend
-	b, backendDiags := c.Backend(&BackendOpts{
-		Config: backendConfig,
-	})
-	diags = diags.Append(backendDiags)
-	if backendDiags.HasErrors() {
-		c.showDiagnostics(diags)
+	// Build the operation request
+	opReq, opDiags := c.OperationRequest(be, view, args.Operation)
+	diags = diags.Append(opDiags)
+	if diags.HasErrors() {
+		view.Diagnostics(diags)
 		return 1
 	}
 
-	// Emit any diagnostics we've accumulated before we delegate to the
-	// backend, since the backend will handle its own diagnostics internally.
-	c.showDiagnostics(diags)
-	diags = nil
-
-	// Build the operation
-	opReq := c.Operation(b)
-	opReq.ConfigDir = configPath
-	opReq.Destroy = destroy
-	opReq.Hooks = []terraform.Hook{c.uiHook()}
-	opReq.PlanOutPath = outPath
-	opReq.PlanRefresh = refresh
-	opReq.ShowDiagnostics = c.showDiagnostics
-	opReq.Type = backend.OperationTypePlan
-	opReq.View = views.NewOperation(arguments.ViewHuman, c.RunningInAutomation, c.View)
+	// Set the plan-specific operation parameters
+	opReq.Destroy = args.Destroy
```

Comment:
It seems kind of odd to me that these are set here, instead of in c.OperationRequest where the other `opReq` parameters get set. 

Is there a reason you aren't passing the entire `args` structure to `OperationRequest` and doing it all in one place?

Reply:
Yeah, this is a bit of a mess, sorry. I'm trying to pass only the minimum data to the function that it needs. It felt wrong to pass the entire args when there's a field called `Operation` in there which is supposed to have the arguments specific to building an operation.

What it comes down to is that I made a mistake when defining the `arguments.Operation` type. It was intended to encompass everything needed to build an operation request, but I missed these two fields from plan.

My suggestion for fixing this here would be:

- Add `Destroy` and `PlanOutPath` to `arguments.Operation` and use those in `OperationRequest`
- Update `arguments.Plan` accordingly
- In apply, add a comment to the line `op.Req.Destroy = c.Destroy` to explain why we're ignoring the `arguments.Operation.Destroy` field (ugh)

Does that make sense to you?

Reply:
Wait, cancel that: it doesn't make sense. We can't add those fields to the operation struct because it would result in changing the flags available for the other commands using the extended flag set.

Reply:
I do prefer that, yes please and thank you! I know it's not always as satisfying to have longish function parameter lists, but I think keeping the operation parameters together is more important for future developers. (though again, this is already approved and you are free to decide it's fine as it is!)

Reply:
Done! This discussion also helped me fix a bug in which I accidentally added the `-auto-approve` flag to plan and refresh 🙈 Thanks!